File: sdk-as/mamoru-sdk-as/assembly/incident.ts
import { JSONEncoder } from "assemblyscript-json/assembly";
import { encode } from "as-base64/assembly";

export class Incident {
    txHash: string
    severity: IncidentSeverity
    message: string
    data: Uint8Array | null
    address: string

    public constructor(txHash: string, severity: IncidentSeverity, message: string, data: Uint8Array | null = null, address: string = "") {
        this.txHash = txHash;
        this.severity = severity;
        this.message = message;
        this.data = data;
        this.address = address;
    }

    toJSON(): string {
        let encoder = new JSONEncoder();

        {
            encoder.pushObject(null);
            encoder.setString("severity", incidentSeverityToString(this.severity));
            encoder.setString("message", this.message);
            encoder.setString("tx_hash", this.txHash);

            if (this.address != "") {
                encoder.setString("address", this.address);
            }

            const data = this.data;

            if (data != null) {
                encoder.setString("data", encode(data));
            }

            encoder.popObject();
        }

        return encoder.toString();
    }
}

export enum IncidentSeverity {
    Info,
    Warning,
    Error,
    Alert,
}

function incidentSeverityToString(severity: IncidentSeverity): string {
    switch (severity) {
        case IncidentSeverity.Info:
            return "info"
        case IncidentSeverity.Warning:
            return "warning"
        case IncidentSeverity.Error:
            return "error"
        case IncidentSeverity.Alert:
            return "alert"
        default:
            return "UNDEFINED"
    }
}

File: mamoru-core/src/daemon/assembly_script/incident.rs
use serde::{Deserialize, Serialize};
use serde_with::base64::Base64;
use serde_with::serde_as;

/// The incident reported by a Daemon.
#[serde_as]
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct IncidentV1 {
    pub severity: IncidentSeverity,
    pub message: String,

    #[serde(default)]
    pub tx_hash: String,

    #[serde(default)]
    pub address: String,

    #[serde_as(as = "Base64")]
    #[serde(default)]
    pub data: Vec<u8>,
}

impl From<IncidentV1> for crate::Incident {
    fn from(value: IncidentV1) -> Self {
        Self {
            severity: value.severity.into(),
            message: value.message,
            tx_hash: value.tx_hash,
            address: value.address,
            data: value.data,
        }
    }
}

#[derive(Debug, Clone, Serialize, Deserialize, Eq, PartialEq)]
#[serde(rename_all = "snake_case")]
pub enum IncidentSeverity {
    Info,
    Warning,
    Error,
    Alert,
}

impl From<IncidentSeverity> for crate::IncidentSeverity {
    fn from(value: IncidentSeverity) -> Self {
        match value {
            IncidentSeverity::Info => Self::Info,
            IncidentSeverity::Warning => Self::Warning,
            IncidentSeverity::Error => Self::Error,
            IncidentSeverity::Alert => Self::Alert,
        }
    }
}
